### DISCUSSION TITLE
System design - correct approach for costly pod starts

### DISCUSSION THREAD

**Discussion by RonaldGalea** [apache/airflow#33922] [General] [answered]
Hello,

I wish to better understand the certain design decisions that Airflow makes, as well as how I could best make use of it for my use case. Specifically, I'd like to gain insights into *why* the KubernetesExecutor operates the way it does, using pod-per-task semantics, and what the alternatives are/could be.

## My use case

I have several services that I need to chain together in some logical way that fits a DAG, so a Workflow Management tool like Airflow seems to fit my use case well. However, some of these services have a relatively high start-up time (large container images, preparing machine learning models, lots of runtime dependencies, etc.), and for these it would be really painful to tear down the warm environment and recreate it constantly.

KubernetesExecutor would recreate pods from scratch, and even the CeleryExecutor creates separate processes to run tasks, so there is no way to reuse a warm environment - but surely I'm not the only one with a use case where recreating execution environments is expensive? So I have the following questions:

- Why does Airflow not support something like pod worker pools that sit behind e.g. a message queue? Is this just not common enough to make it worth the implementation complexity or is there some **fundamental anti-pattern** that makes it unfeasible? If the latter is the case, I would really like to learn more about this case.
- What is the correct system design in the case where some services are too expensive to recreate the environment for? One idea that comes to mind is to replace the actual service in the workflow with a "shell"/"dummy" task that calls the actual service which is backed by a pool of workers elsewhere. However, I'm not sure if such gymnastics are a clean system design.

I would be very thankful for insights regarding the above highlighted considerations.

## Tradeoffs: pod-per-task vs pod worker pool

In this section I wish to highlight some pros and cons for each approach, these are just my thoughts stemming from attempts to explain why Airflow (and other workflow management systems) mainly operate with a pod-per-task approach.

### Pod per task

**Pros:**

- simple implementation
- simple monitoring & error handling (watch what happens to the pod running the task)
- strong isolation (no two tasks can interfere with each other)
- very generic (a task just needs a container image and command to execute)

**Cons:**

- start-up overheads: the pod start up time must be paid for each and every task

### Pod worker pool

To keep things simple, let's assume single-threaded workers, so task concurrency per worker is just 1.

**Pros:**

- warm instances are available, smaller latencies, better scalability

**Cons:**

- complex implementation
- monitoring and error handling become more complicated (clients probably will have to have some logic to support this)
- though less likely when concurrency=1, tasks could still interfere with one another in some cases
- harder to make things generic, one might need to implement clients in multiple languages.

Edit: The idea of a "warm pool of workers" is very general and well-known, so I find it counter-intuitive that it's just not there for Kubernetes pods. For instance, there could be a lightweight client that listens for requests or messages and calls a user-defined callback. Is there some **fundamental reason/limitation** why this is not done? 

Thank you in advance for contributing ideas to this discussion!

**Comment by potiuk** [comment]
You should run CeleryExecutor or even CeleryKubernetesExecutor for that case. They are precisely foreseen to handle your case.

You can run (and our Helm Chart supports it) those executors on K8S.

* https://airflow.apache.org/docs/apache-airflow/stable/core-concepts/executor/celery.html
* https://airflow.apache.org/docs/apache-airflow/stable/core-concepts/executor/celery_kubernetes.html

With CeleryKubernetesExecutor you can even choose which task you run on Cellery (faster, less isolation, instanteneous startup) and which one you run in isolated new pod. 

This was a deliberate design choice - relying on Celery rather than some custom pre-warmed Pods. It has even the advantage that you can have N parallel task in such single running Celery Pod when you deploy the executor on K8S - which gives even better resource utilization.

I recommend you to look at those instead of thinking about another solution

**Reply by potiuk** [reply]
(and I guess your description and details precisely describe how CeleryExecutor works BTW).

**Comment by potiuk** [comment]
Also - interesting thing. My question is - how come you have not found it and spend so much time on describing what actually the Celery Executor is @RonaldGalea ? 

Celery Executor has been there from the beginning, CeleryK8s - for probably >3 years. We have quite a good description of what they do and how they work and with your pro/con description you nicely described 1-1 what Celery Executor does and your arguments are very sound - they even talk about some "later stage" things like concurrency, messages (we use celery kombu with rabbitmq/ or redis or DB as message broker for that) etc. etc. /

Maybe this means that our documentation needs to be updated? Maybe you would like to turn your "pro/con" proposal into a nice documentation in https://airflow.apache.org/docs/apache-airflow/stable/core-concepts/executor/index.html that could explain people like you what's the difference, explain pros and cons etc.?

I think I would love such a documentation PR which could make it clearer to users what they can do with Airflow - you seem to know what you are talking about, yet somehow you missed the core functionality of Airflow about it so maybe such documentation is needed? 

Maybe that's a good idea to contribute such documentation ?

**Reply by RonaldGalea** [reply]
Dear Jarek,

First, I want to thank you for your reply, it cleared up much of my doubt.

## CeleryExecutor

Regarding the CeleryExecutor, it's not that I've missed it, but I've initially moved away from it because it still [creates a new process](https://airflow.apache.org/docs/apache-airflow/stable/core-concepts/executor/celery.html#task-execution-process) per task - which unfortunately is still very expensive in my case.

However, seeing your reply, it's now clear that the issue with my system actually is this expensive process launch, since Celery is meant to handle the fast inference cases - these should probably not have high start-up times, or if they fundamentally do, they really need to be handled with some custom solution, would you agree?

Initially I was looking at some sort of lightweight client that does thread-per-task, but I see how this is difficult & cumbersome (language specific, hard to monitor/handle errors, etc.)

## Documentation

Regarding documentation, I would be very happy to contribute if you deem my feedback helpful. The description of the CeleryExecutor is great, but there are some other parts of the docs that I felt could be improved, not necessarily regarding _how_ things are done, but _why_ they are done the way they are done. This sort of insight enables me to understand the system more deeply, use it efficiently and reason about it to colleagues. I list a couple of examples below:

Celery workers [must have access to the DAGS folder](https://airflow.apache.org/docs/apache-airflow/stable/core-concepts/executor/celery.html#task-execution-process) and it is stated that a common setup would be syncing via some deployment tool, e.g. Chef. I'm sure there is some reason for this, but from my initial impression just from reading this, it seems like an awful lot of complexity for something that is not immediately clear why it's even needed? Why does each individual worker need to know about the _whole_ graph? Is a worker not just supposed to run the task it is given?

Another part I struggled with was understanding what the [Airflow Docker Image](https://airflow.apache.org/docs/docker-stack/index.html) is meant for - what is the concept here? I found an example [cluster configuration](https://airflow.apache.org/docs/apache-airflow/stable/docker-compose.yaml) where it seems that this image is common for all airflow components: webserver, scheduler, worker, etc. Seeing this I can infer that this likely is a strategy to simplify deployment by having a single image - but it would be very helpful to make this intention clear. The [page](https://airflow.apache.org/docs/docker-stack/build.html) describing the customization process of this image is extremely detailed (which is nice), but I couldn't find anything explaining the reasoning behind this image itself, or what it represents conceptually. I also found it counter-intuitive that all workers (even [KubernetesExcutor workers](https://airflow.apache.org/docs/apache-airflow/stable/core-concepts/executor/kubernetes.html#base-image)) must be based on this main Airflow image - I had expected at least that the workers may install a relatively lightweight client such that I can reuse my existing images and just change the entrypoint maybe, is this possible?

Thank you again for your help, and I'm looking forward to potentially improving the docs with the insights from this discussion!

**Comment by RonaldGalea** [comment]
Seems like I did miss a key piece of information, namely that the Celery concurrency model is actually [configurable](https://airflow.apache.org/docs/apache-airflow-providers-celery/stable/configurations-ref.html#pool), which means executing tasks "inline" in a single thread is actually possible with the `solo` option - this should be the fit to my use case

**Reply by potiuk** [reply]
> However, seeing your reply, it's now clear that the issue with my system actually is this expensive process launch, since Celery is meant to handle the fast inference cases - these should probably not have high start-up times, or if they fundamentally do, they really need to be handled with some custom solution, would you agree?

Actually you did not look deep enough. Celery executor by default uses "prefork" which is FAR cheaper than creating new process. You migh not realise that, but prefork is not expensive at all. It's as little expensive as it can be. Yes. It will create a new process, but this process will be already (using your terminology) "warmed up".  It's as good as it can and basically you cannot do better, when you want the "warmed" processes to be somewhat isolated.

Forking on a POSIX-compliant system works in the way that it spawns a new process starting from the place where the parent process forks out. And while doing it, it retrains all the heap memory that the original parent process had. The memory segments for the child, forked process is marked as "copy on write" which means that it takes as minimum resources as possible. Even if the new process reports a lot of memory used, the memory is actually shared with the parent process.

So the new "forked" process takes very small time to start and takes no extra memory when it is forked.  By going `solo` you actually make your tasks less isolated and you can have potential side efects from one task leaking memory or data, it's also potantially security leaks as tasks can potentially easily see data stored in memory by the previous.

> Why does each individual worker need to know about the whole graph? Is a worker not just supposed to run the task it is given?

Because single python code can import another python code (utitlities and commands) and even read other (non-python) files from the folder. Ffor example json file describing how to build your graph in a dynamic way based on metadata stored there). This is why minimum "unit" of sharing is DAG folder - and this is what our users can rely on - that at any point in time they can use and read any python or non-python code in the entire DAG folder. When you start using more of airlfow, learn about dynamic DAGs , dynamic tasks and you wil be in the environment where you have common code maintained by someone else and individual DAGs using that code, you wil understand the power of it. It's hard to see when you focus to only single narrow case, but if you take into accoun that Airlfow is a product and has 100s of thousands installattion all of the world varying from few dags to 10s of thousands of DAGs you will see what it's there.

Of course we could try to prepare and package only the task, it's dependent python code and files that is needed and send it to the task for execution - it would require - every time - re-parsing the files before task is started extracting what is needed, packaging, sending to the remote task and let the task to unpack re-parse and start executing it... But wait... That would totally undermine your case when you want a very quickl start and expect "warmed" worker. Taking into account that each worker can start any task, the fact that Airflow architecture requires pre-syncing and pre-sending all the DAG files to all the workers is actually the reason why your case can be hanled. So you should actually be praising that decision, otherwise the case of your or nearly instantenious start of a task would be next to impossible.

> what the [Airflow Docker Image](https://airflow.apache.org/docs/docker-stack/index.html) is meant for -

As described in the first [page of the docs](https://airflow.apache.org/docs/docker-stack/index.html) - airflow is a reference image, That allows poeople to start and use it. But you are free to customize it to as small or as big as you want. You can even start with SLIM image (possibly you missed that when you read the page) that is targetted for you to start with if you feel comfortable with extending the image only contains 'airflow" core and nothing else and you can extend it as you want. But if you would like to describe it better - feel free to provide the docuemntation :). We are all volunteers here, and you might become one if you think the purpose of the image can be better described. Just take into account that there are various users. Not everyone has the same assumptions and experience as you. Many of our users use the reference image as-is (as described it contains commonly used providers and libraries) becaue they have no idea and they run away from building their image. It would be a pity to miss on those tens of thousands of users.

> This likely is a strategy to simplify deployment by having a single image - but it would be very helpful to make this intention clear

Not reeally it is that wev'e only recently added mechanism that could potentially allow to use different images. Scheduler also needs to parse files to perform scheduling. It needs to parse the files. Which means that if those files import some external libraries. Similarly webserver is able to instantiate certain classes from operators and Hooks that would require to import dependencies of those. Only recently we disabled certain feautres of webserver (connection testing for example) or allowed to run a "standarlone dag processor" that would remove the need of importing those DAG files, Operators, Hook. And by all means we plan to further optimize things in this direction.

>  the workers may install a relatively lightweight client such that I can reuse my existing images and just change the entrypoint maybe, is this possible?

No.   Each worker need at least the providers and librarires they depend on - so that image cannot be "smaller" - as explained above every worker can run any task so it might need ANY dependency. And for the speed reasons you are so concerned with, those dependencies are better to be there installed in the image and "pre-warmed" if you want to start any of your tasks instantnly. We could of course build dynamically such images or extend them dynamically but the speed will suffer and complexity wil grow.

Looking forward to your documentation updates. I hope my explanations will dispelll some of the assumptions you had - without knowing the full picture. I hope you can help us to describe it in the way so that people like you find it crystal clear immediately when they read the docs.

**Reply by RonaldGalea** [reply]
Thank you very much again for your answer.

I still need to make sure I correctly understand the Celery worker setup.

Please let me reiterate my use case: I have several microservices with wildly different hardware requirements and software dependencies.

A single "all-executing" worker was never my intention - rather multiple celery workers listening on specific [queues](https://airflow.apache.org/docs/apache-airflow/stable/core-concepts/executor/celery.html#queues) and the DAG directing tasks to the specific workers.

It is in this sense that I mean workers don't need to know about the whole thing - neither in terms of software dependencies nor the entire graph. Though to the best of my understanding, Airflow expects everything to be defined in a shared, central DAGS_FOLDER for [scheduling](https://airflow.apache.org/docs/apache-airflow/stable/authoring-and-scheduling/dagfile-processing.html#). For instance, there is currently no such feature as an "abstract task", where the DAG only has a task/function interface and the worker actually implements it (similar to [Celery signatures](https://docs.celeryq.dev/en/stable/userguide/canvas.html#signatures)).

If I cannot have such task-specific workers, it means I'd be centralizing everything (codebase & software dependencies) into a single place - unifying the microservices into a monolithic codebase.

Perhaps I misunderstood Airflow conceptually and this is not a valid way to use it? Or rather the features supporting such a use case are not yet implemented? I tend to believe the latter since tasks can already be sent to specific queues, it's just that the workers seem to still need full knowledge about all graphs (cannot be meant for a single task), as well as the scheduler/DAG processor knowing the full implementation of the graphs (no abstract tasks). There already seems to be some effort in this [direction](https://airflow.apache.org/docs/apache-airflow/stable/best-practices.html#using-multiple-docker-images-and-celery-queues).

Thank you again. I will try to put together documentation-related PRs this weekend.

**Reply by potiuk** [reply]
> For instance, there is currently no such feature as an "abstract task", where the DAG only has a task/function interface and the worker actually implements it (similar to [Celery signatures](https://docs.celeryq.dev/en/stable/userguide/canvas.html#signatures)).

The operator concept and Task Flow is the "abstract task". Basically what Celery does (or any other executor) is that it executes `execute` method of the class deriving from BaseOperator (plus a few more things such as JINJA template parameters pre-processing and xcom-push of return value. Task Flow is basically a syntactic sugar around it allowing you to decorate any function with the `@task` converting it to such BaseOperator-derived class on the flight.

> If I cannot have such task-specific workers, it means I'd be centralizing everything (codebase & software dependencies) into a single place - unifying the microservices into a monolithic codebase.

Yes. they can. You can run celery worker running as handling specific queue and mark your task as requiring the queue by `queue` parameter. Then celery tasks will be picking only tasks which are assigned to specific queue. Not sure about your parallel about monolithic vs. microservice though - it's a very long way from one to the other and I fail to connect how one has anything to do with the other.  It's not the matter of "knowledge" af the whole graph and tasks, it's the matter of convenience. Celery worker will not have to "know" the whole list of DAGs if it will only use some parts of the DAG folder tree. And if you wish and are absolutely certain your celery worker will run from specific queue and that all the files, imports etc. required for all the tasks belonging to that folder, then yes - your deployment option might be to only synchronize that folder and nothing else. But currently there are no mechanisms to enforce it - it's up to the DAG author to define which queue each task belong to. You can "simulate" it by implementing custom dag policies https://airflow.apache.org/docs/apache-airflow/stable/administration-and-deployment/cluster-policies.html#dag-policies so you could potentially "enforce" setting "specific" queue for all dags belonging to a specific folder.. So this is generally possible even now, but requires a lot of customizations.

>  Or rather the features supporting such a use case are not yet implemented?

Yes. That's good assumption. We are working on broadly-spaking "Multi-tenancy" - which has several AIPs (Airflow Improvement Proposals). I am leading it. It's not a high priority but we slowly implement something that we discussed at several meetings (you can see the histry and even recordings. https://cwiki.apache.org/confluence/pages/viewpage.action?pageId=89066609 

The "done" AIP is 

AIP-43: [AIP-43 DAG Processor separation](https://cwiki.apache.org/confluence/display/AIRFLOW/AIP-43+DAG+Processor+separation) - where we introduced option of sparate DAG file processor that can only scan sub-folders of DAG folder.

The current work in progress are:

AIP-44:  [AIP-44 Airflow Internal API](https://cwiki.apache.org/confluence/display/AIRFLOW/AIP-44+Airflow+Internal+API)
AIP-56: [AIP-56 Extensible user management](https://cwiki.apache.org/confluence/display/AIRFLOW/AIP-56+Extensible+user+management)

There is yet another AIP to be written (albeit a rather small one) and we are indeed basing the rest of the implementation on a  `sub-folder` == `queue` == `tenant`  and the  AIP wil just make sure that remaining components (triggerer, internal-api, worker) will work `per-queue`  and that dag file processor will be able to "force" queue for all tasks it processes, effectively making "sub-folder" = "queue" = "tenant" , So you will have the option to have a different environment "per sub-folder" or DAG - including different set of dependencies, physical isolation, no access to the DAGs, connecions,variables of other tenants.

It's not been yet proposed formally - we (the three people mostly involved in multi-tenancy implementation) are going to announce it in 2 weeks at the Airflow Summit at this talk https://airflowsummit.org/sessions/2023/multi-tenancy-state-of-the-union/ - it's pending community discussion and approval but I am hopefull we can convince others, get it approved and implemented. We are likely to get discussion going before as well.

Once this will be implemented, you will be able to split Airflow in several parallel "tenants" (except scheduler and webserver that will be common , and shared database but isolated between the tenants via internal API).  This should be "easier" than now - however everyone who would like to use it will have to setup their own tenant deployment.

BTW. If you wish, you can of course take part in the discussions that will happen in the devlist, also you could take part in implementing it if you feel like it.

**Reply by RonaldGalea** [reply]
Here's a more concrete example showing my original idea related to "abstract tasks"/"signatures".

Assume I have microservice A and B, separate repositories, separate code and environment (software dependencies).

Both microservices expose a simple API (just some functions): `do_job_A(args_A) -> output_A` and `do_job_B(args_B) -> output_B`.

## Original idea

I would want to be able to do the following (I introduce the `@signature` decorator as a hypothetical mechanism to define the graph):

```py
from airflow.decorators import dag, signature
@dag(
    # parameters...
)
def example_dag():
    # ... get some args from external trigger
    # args_for_task_A = ...

    @signature(queue='queueA')
    def do_job_A(args_A) -> output_A:
        pass

    @signature(queue='queueB')
    def do_job_B(args_B) -> output_B:
        pass

    ... = do_job_A(args_for_task_A)
    ... = do_job_B(...) # maybe use the output of A...
    # continue processing B_output...

example_dag()
```

Notice that this graph definition file is independent of any microservice code/dependencies. It is a pure graph definition. What `@signature` does is the following:

- serializes inputs
- sends serialized inputs and `task_id` to the right queue (where the worker has the concrete implementation of the task)

This is pretty much exactly what [Celery signatures](https://docs.celeryq.dev/en/stable/userguide/canvas.html#signatures) are designed to do (for reference, consider the following small example I've set up: there is a small Celery worker that defines [my_dummy_task](https://github.com/RonaldGalea/dummy-celery-worker/blob/main/tasks.py#L12). In a separate repo, there is a small client [trigerring my_dummy_task](https://github.com/RonaldGalea/dummy-celery-client/blob/main/other_codebase_client.py#L6) via a signature - not knowing any implementation details/dependencies of the task), and with setup like the above, the graph definition remains really simple and its purpose is clear: just direct tasks to workers.

I'm guessing that the the closest option to this is the ExternalPythonOperator.

## Perhaps with ExternalPythonOperator?

For example:

```py
from airflow.decorators import dag, task

@dag(
    # parameters...
)
def example_dag():

    @task.external_python(queue='queueA')
    def do_job_A_task(args_A) -> output_A:
        from my_package_A import do_job_A
        return do_job_A(args_A)

    @task.external_python(queue='queueB')
    def do_job_B_task(args_B) -> output_B:
        from my_package_B import do_job_B
        return do_job_B(args_B)

    ... = do_job_A_task(...)
    do_job_B_task(...)
```

This is a bit more cumbersome as I have to make custom packages and still share the DAGS_FOLDER with all workers, even though I'd only be sending specific tasks to specific queues. Still, it is be close enough and should work.

Many thanks again for the helpful discussions.

**Reply by potiuk** [reply, marked as answer]
Your example of signatures would work - but only under the assumption that your tasks do not use any other imports - specifically imports from other files in the DAG folder. While nicely academic "encapsulation", in practice it's not how most users in scenarios whre users are running in corporate environments and multiple teams working on the data pipelines the architecture of their DAGs will - sooner or later - start to fall under [Conway's law](https://en.wikipedia.org/wiki/Conway%27s_law) . In short the structure of their DAGs packages will start to look like organisational structure of their teams..

You can see typical (simplified) organisational structure that many companies have

https://airflow.apache.org/docs/apache-airflow/stable/administration-and-deployment/modules_management.html#typical-structure-of-packages

In this structure - you will have common code extracted to some common packages, and many of those tasks that you think are nicely "isolated" and can be just serialized and passed for remote execution, will in fact start using the common code. And tha common code will start using common code written by other teams as well. and so on. In fact if you would like to find out which code should be serialized and sent, you will have to parse the whole tree and (essentially) send not only the task - but also alll the imports that are done by the tasks from the DAG folder (assuming that we make sure execution environment has the same packages installed as when you parse the DAG). The imports could be dynamic. The Python tasks might even read locally (in DAG) stored files with meta-data that could be used to run the task. So effectively you would have to figure out which "slice" of the DAG folder you'd need to run your task on a remote system with and you'd have to "transplant" that slice to remote worker.

Of course you could attempt to limit som of those operations, but then it would be a subset of Python, not Python and it would be extremelly difficult to sand-box it in the way that those operations woudl not be allowed. I believe Python core team tried implementing sandboxed Python years ago and all the attempts failed miserably due to the very dynamic nature of Python.

The assumption of Airlfow is that syncing the whole DAG folder and telling the user "you can use anything in the DAG folder in your task" is considerably simpler, and in vast majority of cases even more performant and faster, because same tasks or nodes where the DAGs get synchronized might be effectively used to run multiple tasks and there is the effect of sending the whole DAG folder once when you run thousands of various tasks on the same worker/node.  If you want to "isolate" the task and serialize only what it needs to execute it every time, this will be far more to transfer, unless you will do some caching which will keep the transferred data on the target worker... But if you keep the cache on the worker eventually you will end up with the whole DAG folder in the cache on the worker.  Whic makes 360 circle because we end up with whole DAG folder synchronized on all the workers, only using super complex mechanism of slicing and caching, rather than syncing the whole DAG folder in the first place - once.

So effectively the "practical" implementation of your idea despite it's "academic elegance" would be IMHO:

1) terriblly complex to implement (thus prone to errors)
2) less performant than syncing the whole DAG folder
3) ending up with the same "whole DAG anyhow on the workers"
